### ISSUE TITLE
re-initialize pip master working set after installs

### ISSUE THREAD

**Issue by kafonek** [nbgallery/ipydeps#15]
When installing packages with dependencies on higher versions of the idna package, ipydeps will fail with a VersionConflict Exception.  For instance, `ipydeps.pip('requests')` will install (as of 15 June 2017) `certifi-2017.4.17, chardet-3.0.4, idna-2.5, requests-2.17.3, urllib3-1.21.1`.  A second call to `ipydeps.pip('requests')` or to any library that requires requests as a dependency will raise a VersionConflict saying that requests needs idna >=2.5 but it found idna 2.2 on the system.

One workaround is to explicitly call [pip.utils.pkg_resources._initialize_master_working_set()](https://github.com/pypa/pip/blob/master/pip/_vendor/pkg_resources/__init__.py#L3038) in between `ipydeps.pip` calls.  For instance this will work - 

```python
import ipydeps
import pip
ipydeps.pip('requests') # installs certifi, chardet, idna-2.5, etc
pip.utils.pkg_resources.get_distribution('idna') # this will show idna 2.2

### if we ran ipydeps.pip('requests') again it would raise VersionConflict
pip.utils.pkg_resources._initialize_master_working_set()
pip.utils.pkg_resources.get_distribution('idna') # this will show idna 2.5 now.

ipydeps.pip('requests') # works
```

Pip documentation mentions "Invocation [of _initialize_master_working_set] by other packages is unsupported and done at their own risk."  I do not know what side effects this might have.  Fixing the VersionConflict exception is very important though.

**Comment by billallen256** [comment]
Just to document what I tried today...
- The issue appears to manifest itself when ipydeps.pip is called for the requests package (or something that depends on the requests package) twice in the same notebook
- I tested out adding pkg_resources._initialize_master_working_set() right after the call to _invalidate_cache()
- There was no effect.  pip.get_installed_distributions() continued to report idna 2.2 while pkg_resources.get_distribution('idna') reported idna 2.5
- Started exploring custom loaders in importlib and the built in package called "site"

**Comment by jeffyjefflabs** [comment]
Possible the [patch](https://github.com/nbgallery/jupyter-docker/commit/b878942d25ecbc658673112a03840c0ba39b3f49) in #7 is connected?  Not sure that patch would *cause* this, but a fix here might also fix that.

**Comment by billallen256** [comment]
Turns out that it's really important to use the vendored pkg_resources.  Had much better luck messing with this on Friday.  Please roll in the following changes for testing:
https://github.com/nbgallery/ipydeps/commit/ff822ed50fd33fe66ffb1877cc79f922d571bac7

**Comment by billallen256** [comment]
Resolved in db7ea45
